gamla/functional_generic.py

```diff
@@ -338,5 +340,55 @@ def bifurcate(*funcs):
 
 def value_to_dict(key: Text):
     return compose_left(
-        functional.wrap_tuple, functional.prefix(key), functional.wrap_tuple, dict,
+        functional.wrap_tuple,
+        functional.prefix(key),
+        functional.wrap_tuple,
+        dict,
     )
+
+
+_R = TypeVar("_R")
+_E = TypeVar("_E")
+
+
+def reduce_curried(
+    reducer: Callable[[_R, _E], _R],
+    initial_value: _R,
+) -> Callable[[Iterable[_E]], _R]:
+    if asyncio.iscoroutinefunction(reducer):
+
+        async def reduce_async(elements):
+            state = initial_value
+            for element in elements:
+                state = await reducer(state, element)
+            return state
+
+        return reduce_async
+
+    def reduce(elements):
+        state = initial_value
+        for element in elements:
+            state = reducer(state, element)
+        return state
+
+    return reduce
+
+
+@toolz.curry
```

Comment:
why not `gamla.curry`?

Reply:
Everything here should be `gamla.curry`, I want to do it all at once in a followup commit.